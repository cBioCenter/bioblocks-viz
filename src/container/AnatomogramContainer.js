"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// tslint:disable-next-line:import-name
var anatomogram_1 = require("anatomogram");
var immutable_1 = require("immutable");
var React = require("react");
var react_redux_1 = require("react-redux");
var redux_1 = require("redux");
var action_1 = require("~bioblocks-viz~/action");
var component_1 = require("~bioblocks-viz~/component");
var container_1 = require("~bioblocks-viz~/container");
var data_1 = require("~bioblocks-viz~/data");
var helper_1 = require("~bioblocks-viz~/helper");
var reducer_1 = require("~bioblocks-viz~/reducer");
var selector_1 = require("~bioblocks-viz~/selector");
var AnatomogramContainerClass = /** @class */ (function (_super) {
    tslib_1.__extends(AnatomogramContainerClass, _super);
    function AnatomogramContainerClass(props) {
        var _this = _super.call(this, props) || this;
        _this.divRef = null;
        _this.onClick = function (ids) {
            // Anatomogram returns an array of strings for click events, but we only ever work with a single id.
            var id = ids[0];
            var _a = _this.props, addLabel = _a.onLabelAdd, removeLabel = _a.onLabelRemove, selectIds = _a.selectIds;
            if (selectIds.includes(id)) {
                removeLabel(id);
            }
            else {
                addLabel(id);
            }
        };
        _this.deriveIdsFromSpecies = function (species) { return Object.keys(data_1.AnatomogramMapping[species]); };
        _this.onMouseOut = function (id) {
            return;
        };
        _this.onMouseOver = function (id) {
            return;
        };
        _this.parseCategory = function (category) {
            var splitCategories = category.split(/-|_/);
            return splitCategories[0];
        };
        _this.resizeSVGElement = function (error, svgDomNode) {
            if (_this.divRef) {
                var isSvgHeightBigger = svgDomNode.height.baseVal.value > svgDomNode.width.baseVal.value;
                // The Anatomogram Component internally sets the svg height to 'auto'.
                // So, to allow more flexibility in sizing it, we have to manually override it here. Sorry.
                svgDomNode.style.height = isSvgHeightBigger ? "calc(" + _this.divRef.style.height + " - 50px)" : 'auto';
                svgDomNode.style.padding = '0';
                svgDomNode.style.width = isSvgHeightBigger ? 'auto' : "calc(" + _this.divRef.style.width + " - 75px)";
            }
        };
        _this.state = {
            ids: _this.deriveIdsFromSpecies(props.species),
        };
        return _this;
    }
    AnatomogramContainerClass.prototype.setupDataServices = function () {
        this.registerDataset('cells', []);
        this.registerDataset('labels', []);
        reducer_1.BioblocksMiddlewareTransformer.addTransform(this.getCellToLabelTransform());
        reducer_1.BioblocksMiddlewareTransformer.addTransform(this.getLabelToCellTransform());
    };
    AnatomogramContainerClass.prototype.componentDidUpdate = function (prevProps) {
        var species = this.props.species;
        if (species !== prevProps.species) {
            this.setState({
                ids: this.deriveIdsFromSpecies(species),
            });
        }
    };
    AnatomogramContainerClass.prototype.render = function () {
        var _this = this;
        var _a = this.props, iconSrc = _a.iconSrc, species = _a.species, selectIds = _a.selectIds;
        var ids = this.state.ids;
        return (React.createElement(react_redux_1.Provider, { store: reducer_1.BBStore },
            React.createElement("div", { className: 'anatomogram-container', ref: function (node) {
                    if (node) {
                        _this.divRef = node.getElementsByTagName('div')[0];
                    }
                } },
                React.createElement(component_1.ComponentCard, { componentName: 'Anatomogram', iconSrc: iconSrc },
                    React.createElement(anatomogram_1.default, { atlasUrl: "", highlightColour: 'yellow', onClick: this.onClick, onInjected: this.resizeSVGElement, onMouseOut: this.onMouseOut, onMouseOver: this.onMouseOver, selectColour: 'ffaa00', selectIds: selectIds.toArray(), showIds: ids, species: species, selectedView: species === 'mus_musculus' ? 'female' : 'male' })))));
    };
    AnatomogramContainerClass.prototype.getCellToLabelTransform = function () {
        var _this = this;
        return {
            fn: function (state) {
                var e_1, _a, e_2, _b, e_3, _c;
                var species = _this.props.species;
                var currentCells = selector_1.selectCurrentItems(state, 'cells').toArray();
                var _d = selector_1.getSpring(state), category = _d.category, graphData = _d.graphData;
                var result = immutable_1.Set();
                try {
                    for (var currentCells_1 = tslib_1.__values(currentCells), currentCells_1_1 = currentCells_1.next(); !currentCells_1_1.done; currentCells_1_1 = currentCells_1.next()) {
                        var cellIndex = currentCells_1_1.value;
                        // Parse labels and categories from SPRING data.
                        var labelForCategory = graphData.nodes[cellIndex] ? graphData.nodes[cellIndex].labelForCategory : {};
                        var labels = data_1.AnatomogramMapping[species][labelForCategory[category]];
                        if (labels) {
                            labels.forEach(function (label) { return (result = result.add(label)); });
                        }
                        try {
                            for (var _e = (e_2 = void 0, tslib_1.__values(Object.keys(data_1.AnatomogramMapping[species]))), _f = _e.next(); !_f.done; _f = _e.next()) {
                                var id = _f.value;
                                try {
                                    for (var _g = (e_3 = void 0, tslib_1.__values(Object.values(labelForCategory))), _h = _g.next(); !_h.done; _h = _g.next()) {
                                        var label = _h.value;
                                        if (data_1.AnatomogramMapping[species][id].includes(label)) {
                                            result = result.add(id);
                                        }
                                    }
                                }
                                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                                finally {
                                    try {
                                        if (_h && !_h.done && (_c = _g.return)) _c.call(_g);
                                    }
                                    finally { if (e_3) throw e_3.error; }
                                }
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (currentCells_1_1 && !currentCells_1_1.done && (_a = currentCells_1.return)) _a.call(currentCells_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return result;
            },
            fromState: { stateName: 'cells' },
            toState: { stateName: 'labels' },
        };
    };
    AnatomogramContainerClass.prototype.getLabelToCellTransform = function () {
        var _this = this;
        return {
            fn: function (state) {
                var _a = _this.props, species = _a.species, selectIds = _a.selectIds;
                var graphData = selector_1.getSpring(state).graphData;
                var anatomogramMap = data_1.AnatomogramMapping[species];
                var candidateLabels = immutable_1.Set();
                selectIds.forEach(function (id) {
                    candidateLabels = id && anatomogramMap[id] ? candidateLabels.merge(anatomogramMap[id]) : candidateLabels;
                });
                var cellIndices = immutable_1.Set();
                graphData.nodes.forEach(function (node) {
                    var labelsForNode = Object.values(node.labelForCategory);
                    candidateLabels.forEach(function (label) {
                        if (label && labelsForNode.includes(label)) {
                            cellIndices = cellIndices.add(node.number);
                            return;
                        }
                    });
                });
                return cellIndices;
            },
            fromState: 'bioblocks/labels',
            toState: 'bioblocks/cells',
        };
    };
    AnatomogramContainerClass.defaultProps = {
        iconSrc: '',
        onLabelAdd: helper_1.EMPTY_FUNCTION,
        onLabelRemove: helper_1.EMPTY_FUNCTION,
        selectIds: immutable_1.Set(),
        species: 'homo_sapiens',
    };
    AnatomogramContainerClass.displayName = 'Anatomogram';
    return AnatomogramContainerClass;
}(container_1.BioblocksVisualization));
exports.AnatomogramContainerClass = AnatomogramContainerClass;
var mapStateToProps = function (state) { return ({
    selectIds: selector_1.selectCurrentItems(state, 'labels'),
}); };
var mapDispatchToProps = function (dispatch) {
    return redux_1.bindActionCreators({
        onLabelAdd: action_1.createContainerActions('labels').add,
        onLabelRemove: action_1.createContainerActions('labels').remove,
    }, dispatch);
};
exports.AnatomogramContainer = react_redux_1.connect(mapStateToProps, mapDispatchToProps)(AnatomogramContainerClass);
//# sourceMappingURL=AnatomogramContainer.js.map