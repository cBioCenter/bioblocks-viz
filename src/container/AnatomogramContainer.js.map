{"version":3,"file":"AnatomogramContainer.js","sourceRoot":"","sources":["../../../src/container/AnatomogramContainer.tsx"],"names":[],"mappings":";;;AAAA,uCAAuC;AACvC,2CAAsC;AACtC,uCAAgC;AAChC,6BAA+B;AAC/B,2CAAgD;AAChD,+BAAqD;AAErD,iDAAgE;AAChE,uDAA0D;AAC1D,uDAAmE;AACnE,6CAAwE;AACxE,iDAAwD;AACxD,mDAAuH;AACvH,qDAAyE;AAoBzE;IAA+C,qDAG9C;IAaC,mCAAY,KAAiC;QAA7C,YACE,kBAAM,KAAK,CAAC,SAIb;QAPS,YAAM,GAA0B,IAAI,CAAC;QA0HrC,aAAO,GAAG,UAAC,GAAa;YAChC,oGAAoG;YACpG,IAAM,EAAE,GAAG,GAAG,CAAC,CAAC,CAAC,CAAC;YACZ,IAAA,gBAA4E,EAA1E,wBAAoB,EAAE,8BAA0B,EAAE,wBAAwB,CAAC;YAEnF,IAAI,SAAS,CAAC,QAAQ,CAAC,EAAE,CAAC,EAAE;gBAC1B,WAAW,CAAC,EAAE,CAAC,CAAC;aACjB;iBAAM;gBACL,QAAQ,CAAC,EAAE,CAAC,CAAC;aACd;QACH,CAAC,CAAC;QAEQ,0BAAoB,GAAG,UAAC,OAAqB,IAAK,OAAA,MAAM,CAAC,IAAI,CAAC,yBAAkB,CAAC,OAAO,CAAC,CAAC,EAAxC,CAAwC,CAAC;QAE3F,gBAAU,GAAG,UAAC,EAAU;YAChC,OAAO;QACT,CAAC,CAAC;QAEQ,iBAAW,GAAG,UAAC,EAAU;YACjC,OAAO;QACT,CAAC,CAAC;QAEQ,mBAAa,GAAG,UAAC,QAAgB;YACzC,IAAM,eAAe,GAAG,QAAQ,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC;YAE9C,OAAO,eAAe,CAAC,CAAC,CAAC,CAAC;QAC5B,CAAC,CAAC;QAEQ,sBAAgB,GAAG,UAAC,KAAU,EAAE,UAAyB;YACjE,IAAI,KAAI,CAAC,MAAM,EAAE;gBACf,IAAM,iBAAiB,GAAG,UAAU,CAAC,MAAM,CAAC,OAAO,CAAC,KAAK,GAAG,UAAU,CAAC,KAAK,CAAC,OAAO,CAAC,KAAK,CAAC;gBAE3F,sEAAsE;gBACtE,2FAA2F;gBAC3F,UAAU,CAAC,KAAK,CAAC,MAAM,GAAG,iBAAiB,CAAC,CAAC,CAAC,UAAQ,KAAI,CAAC,MAAM,CAAC,KAAK,CAAC,MAAM,aAAU,CAAC,CAAC,CAAC,MAAM,CAAC;gBAClG,UAAU,CAAC,KAAK,CAAC,OAAO,GAAG,GAAG,CAAC;gBAC/B,UAAU,CAAC,KAAK,CAAC,KAAK,GAAG,iBAAiB,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC,CAAC,UAAQ,KAAI,CAAC,MAAM,CAAC,KAAK,CAAC,KAAK,aAAU,CAAC;aACjG;QACH,CAAC,CAAC;QA5JA,KAAI,CAAC,KAAK,GAAG;YACX,GAAG,EAAE,KAAI,CAAC,oBAAoB,CAAC,KAAK,CAAC,OAAO,CAAC;SAC9C,CAAC;;IACJ,CAAC;IAEM,qDAAiB,GAAxB;QACE,IAAI,CAAC,eAAe,CAAC,OAAO,EAAE,EAAE,CAAC,CAAC;QAClC,IAAI,CAAC,eAAe,CAAC,QAAQ,EAAE,EAAE,CAAC,CAAC;QACnC,wCAA8B,CAAC,YAAY,CAAC,IAAI,CAAC,uBAAuB,EAAE,CAAC,CAAC;QAC5E,wCAA8B,CAAC,YAAY,CAAC,IAAI,CAAC,uBAAuB,EAAE,CAAC,CAAC;IAC9E,CAAC;IAEM,sDAAkB,GAAzB,UAA0B,SAAqC;QACrD,IAAA,4BAAO,CAAgB;QAC/B,IAAI,OAAO,KAAK,SAAS,CAAC,OAAO,EAAE;YACjC,IAAI,CAAC,QAAQ,CAAC;gBACZ,GAAG,EAAE,IAAI,CAAC,oBAAoB,CAAC,OAAO,CAAC;aACxC,CAAC,CAAC;SACJ;IACH,CAAC;IAEM,0CAAM,GAAb;QAAA,iBAgCC;QA/BO,IAAA,eAA4C,EAA1C,oBAAO,EAAE,oBAAO,EAAE,wBAAwB,CAAC;QAC3C,IAAA,oBAAG,CAAgB;QAE3B,OAAO,CACL,oBAAC,sBAAQ,IAAC,KAAK,EAAE,iBAAO;YACtB,6BACE,SAAS,EAAE,uBAAuB,EAClC,GAAG,EAAE,UAAA,IAAI;oBACP,IAAI,IAAI,EAAE;wBACR,KAAI,CAAC,MAAM,GAAG,IAAI,CAAC,oBAAoB,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAC;qBACnD;gBACH,CAAC;gBAED,oBAAC,yBAAa,IAAC,aAAa,EAAE,aAAa,EAAE,OAAO,EAAE,OAAO;oBAC3D,oBAAC,qBAAW,IACV,QAAQ,EAAE,EAAE,EACZ,eAAe,EAAE,QAAQ,EACzB,OAAO,EAAE,IAAI,CAAC,OAAO,EACrB,UAAU,EAAE,IAAI,CAAC,gBAAgB,EACjC,UAAU,EAAE,IAAI,CAAC,UAAU,EAC3B,WAAW,EAAE,IAAI,CAAC,WAAW,EAC7B,YAAY,EAAE,QAAQ,EACtB,SAAS,EAAE,SAAS,CAAC,OAAO,EAAE,EAC9B,OAAO,EAAE,GAAG,EACZ,OAAO,EAAE,OAAO,EAChB,YAAY,EAAE,OAAO,KAAK,cAAc,CAAC,CAAC,CAAC,QAAQ,CAAC,CAAC,CAAC,MAAM,GAC5D,CACY,CACZ,CACG,CACZ,CAAC;IACJ,CAAC;IAES,2DAAuB,GAAjC;QAAA,iBA+BC;QA9BC,OAAO;YACL,EAAE,EAAE,UAAA,KAAK;;gBACC,IAAA,6BAAO,CAAgB;gBAC/B,IAAM,YAAY,GAAG,6BAAkB,CAAS,KAAK,EAAE,OAAO,CAAC,CAAC,OAAO,EAAE,CAAC;gBACpE,IAAA,gCAA0C,EAAxC,sBAAQ,EAAE,wBAA8B,CAAC;gBACjD,IAAI,MAAM,GAAG,eAAG,EAAU,CAAC;;oBAE3B,KAAwB,IAAA,iBAAA,iBAAA,YAAY,CAAA,0CAAA,oEAAE;wBAAjC,IAAM,SAAS,yBAAA;wBAClB,gDAAgD;wBAChD,IAAM,gBAAgB,GAAG,SAAS,CAAC,KAAK,CAAC,SAAS,CAAC,CAAC,CAAC,CAAC,SAAS,CAAC,KAAK,CAAC,SAAS,CAAC,CAAC,gBAAgB,CAAC,CAAC,CAAC,EAAE,CAAC;wBAEvG,IAAM,MAAM,GAAG,yBAAkB,CAAC,OAAO,CAAC,CAAC,gBAAgB,CAAC,QAAQ,CAAC,CAAC,CAAC;wBACvE,IAAI,MAAM,EAAE;4BACV,MAAM,CAAC,OAAO,CAAC,UAAA,KAAK,IAAI,OAAA,CAAC,MAAM,GAAG,MAAM,CAAC,GAAG,CAAC,KAAK,CAAC,CAAC,EAA5B,CAA4B,CAAC,CAAC;yBACvD;;4BAED,KAAiB,IAAA,oBAAA,iBAAA,MAAM,CAAC,IAAI,CAAC,yBAAkB,CAAC,OAAO,CAAC,CAAC,CAAA,CAAA,gBAAA,4BAAE;gCAAtD,IAAM,EAAE,WAAA;;oCACX,KAAoB,IAAA,oBAAA,iBAAA,MAAM,CAAC,MAAM,CAAC,gBAAgB,CAAC,CAAA,CAAA,gBAAA,4BAAE;wCAAhD,IAAM,KAAK,WAAA;wCACd,IAAI,yBAAkB,CAAC,OAAO,CAAC,CAAC,EAAE,CAAC,CAAC,QAAQ,CAAC,KAAK,CAAC,EAAE;4CACnD,MAAM,GAAG,MAAM,CAAC,GAAG,CAAC,EAAE,CAAC,CAAC;yCACzB;qCACF;;;;;;;;;6BACF;;;;;;;;;qBACF;;;;;;;;;gBAED,OAAO,MAAM,CAAC;YAChB,CAAC;YACD,SAAS,EAAE,EAAE,SAAS,EAAE,OAAO,EAAE;YACjC,OAAO,EAAE,EAAE,SAAS,EAAE,QAAQ,EAAE;SACjC,CAAC;IACJ,CAAC;IAES,2DAAuB,GAAjC;QAAA,iBA4BC;QA3BC,OAAO;YACL,EAAE,EAAE,UAAA,KAAK;gBACD,IAAA,gBAAmC,EAAjC,oBAAO,EAAE,wBAAwB,CAAC;gBAClC,IAAA,iDAAS,CAAsB;gBACvC,IAAM,cAAc,GAAG,yBAAkB,CAAC,OAAO,CAAC,CAAC;gBACnD,IAAI,eAAe,GAAG,eAAG,EAAU,CAAC;gBACpC,SAAS,CAAC,OAAO,CAAC,UAAA,EAAE;oBAClB,eAAe,GAAG,EAAE,IAAI,cAAc,CAAC,EAAE,CAAC,CAAC,CAAC,CAAC,eAAe,CAAC,KAAK,CAAC,cAAc,CAAC,EAAE,CAAC,CAAC,CAAC,CAAC,CAAC,eAAe,CAAC;gBAC3G,CAAC,CAAC,CAAC;gBAEH,IAAI,WAAW,GAAG,eAAG,EAAU,CAAC;gBAChC,SAAS,CAAC,KAAK,CAAC,OAAO,CAAC,UAAA,IAAI;oBAC1B,IAAM,aAAa,GAAG,MAAM,CAAC,MAAM,CAAC,IAAI,CAAC,gBAAgB,CAAC,CAAC;oBAC3D,eAAe,CAAC,OAAO,CAAC,UAAA,KAAK;wBAC3B,IAAI,KAAK,IAAI,aAAa,CAAC,QAAQ,CAAC,KAAK,CAAC,EAAE;4BAC1C,WAAW,GAAG,WAAW,CAAC,GAAG,CAAC,IAAI,CAAC,MAAM,CAAC,CAAC;4BAE3C,OAAO;yBACR;oBACH,CAAC,CAAC,CAAC;gBACL,CAAC,CAAC,CAAC;gBAEH,OAAO,WAAW,CAAC;YACrB,CAAC;YACD,SAAS,EAAE,kBAAkB;YAC7B,OAAO,EAAE,iBAAiB;SAC3B,CAAC;IACJ,CAAC;IAlIa,sCAAY,GAAG;QAC3B,OAAO,EAAE,EAAE;QACX,UAAU,EAAE,uBAAc;QAC1B,aAAa,EAAE,uBAAc;QAC7B,SAAS,EAAE,eAAG,EAAU;QACxB,OAAO,EAAE,cAA8B;KACxC,CAAC;IAEY,qCAAW,GAAG,aAAa,CAAC;IAmK5C,gCAAC;CAAA,AA/KD,CAA+C,kCAAsB,GA+KpE;AA/KY,8DAAyB;AAiLtC,IAAM,eAAe,GAAG,UAAC,KAAgB,IAAK,OAAA,CAAC;IAC7C,SAAS,EAAE,6BAAkB,CAAS,KAAK,EAAE,QAAQ,CAAC;CACvD,CAAC,EAF4C,CAE5C,CAAC;AAEH,IAAM,kBAAkB,GAAG,UAAC,QAAkB;IAC5C,OAAA,0BAAkB,CAChB;QACE,UAAU,EAAE,+BAAsB,CAAS,QAAQ,CAAC,CAAC,GAAG;QACxD,aAAa,EAAE,+BAAsB,CAAS,QAAQ,CAAC,CAAC,MAAM;KAC/D,EACD,QAAQ,CACT;AAND,CAMC,CAAC;AAES,QAAA,oBAAoB,GAAG,qBAAO,CACzC,eAAe,EACf,kBAAkB,CACnB,CAAC,yBAAyB,CAAC,CAAC"}