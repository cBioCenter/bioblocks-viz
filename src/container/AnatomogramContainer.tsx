// tslint:disable-next-line:import-name
import Anatomogram from 'anatomogram';
import { isEqual } from 'lodash';
import * as React from 'react';

import {
  CellContext,
  ICellContext,
  initialCellContext,
  initialSpringContext,
  ISpringContext,
  SpringContext,
} from '~chell-viz~/context';
import { CHELL_CSS_STYLE } from '~chell-viz~/data';

export interface IAnatomogramContainerProps {
  cellContext: ICellContext;
  height: number | string;
  species: 'homo_sapiens' | 'mus_musculus';
  springContext: ISpringContext;
  style: CHELL_CSS_STYLE;
  width: number | string;
}

export interface IAnatomogramContainerState {
  ids: string[];
  selectIds: string[];
}

export const springToAnatomogramMapping: { [key: string]: { [key: string]: string } } = {
  homo_sapiens: {
    P11A: 'UBERON_0000178',
    P11B: 'UBERON_0000473',
    P12A: 'UBERON_0000955',
    P9A: 'UBERON_0000977',
  },
  mus_musculus: {
    bladder: 'UBERON_0001255',
    heart: 'UBERON_0000948',
    kidney: 'UBERON_0002113',
    liver: 'UBERON_0002107',
    lung: 'UBERON_0002048 ',
    mammary: 'UBERON_0001911',
    marrow: 'UBERON_0002371',
    muscle: 'sUBERON_0014892',
    spleen: 'UBERON_0002106',
    thymus: 'UBERON_0002370',
    tongue: 'UBERON_0001723',
    trachea: 'UBERON_0003126',
  },
};

export const anatomogramToSpringMapping: { [key: string]: { [key: string]: string } } = {
  homo_sapiens: {
    UBERON_0000178: 'P11A',
    UBERON_0000473: 'P11B',
    UBERON_0000955: 'P12A',
    UBERON_0000977: 'P9A',
  },
  mus_musculus: {
    UBERON_0000948: 'heart',
    UBERON_0001255: 'bladder',
    UBERON_0001723: 'tongue',
    UBERON_0001911: 'mammary',
    UBERON_0002048: 'lung',
    UBERON_0002106: 'spleen',
    UBERON_0002107: 'liver',
    UBERON_0002113: 'kidney',
    UBERON_0002370: 'thymus',
    UBERON_0002371: 'marrow',
    UBERON_0003126: 'trachea',
    sUBERON_0014892: 'muscle',
  },
};

export class AnatomogramContainerClass extends React.Component<IAnatomogramContainerProps, IAnatomogramContainerState> {
  public static defaultProps = {
    cellContext: {
      ...initialCellContext,
    },
    height: '300px',
    ids: ['UBERON_0000178', 'UBERON_0000473', 'UBERON_0000955', 'UBERON_0000977'],
    springContext: {
      ...initialSpringContext,
    },
    style: {},
    width: '400px',
  };

  public static displayName = 'Anatomogram';

  protected anatomogramRef: any;

  constructor(props: IAnatomogramContainerProps) {
    super(props);
    this.state = {
      ids: Object.keys(anatomogramToSpringMapping[props.species]),
      selectIds: [],
    };
  }

  public componentDidUpdate(prevProps: IAnatomogramContainerProps) {
    const { cellContext, springContext } = this.props;
    if (!isEqual(springContext.selectedCategories, prevProps.springContext.selectedCategories)) {
      this.setState({
        selectIds: springContext.selectedCategories.map(
          category => springToAnatomogramMapping[this.props.species][category],
        ),
      });
    } else if (!isEqual(cellContext.currentCells, prevProps.cellContext.currentCells)) {
      const categories = new Set<string>();
      for (const cellIndex of cellContext.currentCells) {
        categories.add(springContext.graphData.nodes[cellIndex].category);
      }
      this.setState({
        selectIds: Array.from(categories).map(category => springToAnatomogramMapping[this.props.species][category]),
      });
    }
  }

  public render() {
    const { ids, selectIds } = this.state;

    return (
      <div className={'anatomogram-container'} style={{ height: '100%' }}>
        <Anatomogram
          atlasUrl={``}
          highlightColour={'yellow'}
          onClick={this.onClick}
          onMouseOut={this.onMouseOut}
          onMouseOver={this.onMouseOver}
          selectColour={'ffaa00'}
          selectIds={selectIds}
          showIds={ids}
          species={this.props.species}
          selectedView={this.props.species === 'mus_musculus' ? 'female' : 'male'}
        />
      </div>
    );
  }

  protected onClick = (id: string) => {
    this.props.springContext.toggleCategory(anatomogramToSpringMapping[this.props.species][id]);
  };

  protected onMouseOut = (id: string) => {
    return;
  };

  protected onMouseOver = (id: string) => {
    return;
  };
}

type requiredProps = Omit<IAnatomogramContainerProps, keyof typeof AnatomogramContainerClass.defaultProps> &
  Partial<IAnatomogramContainerProps>;

export const AnatomogramContainer = (props: requiredProps) => (
  <CellContext.Consumer>
    {cellContext => (
      <SpringContext.Consumer>
        {springContext => (
          <AnatomogramContainerClass {...props} cellContext={cellContext} springContext={springContext} />
        )}
      </SpringContext.Consumer>
    )}
  </CellContext.Consumer>
);
