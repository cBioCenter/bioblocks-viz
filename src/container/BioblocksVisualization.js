"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var immutable_1 = require("immutable");
var React = require("react");
var util_1 = require("util");
var data_1 = require("~bioblocks-viz~/data");
var reducer_1 = require("~bioblocks-viz~/reducer");
var BioblocksVisualization = /** @class */ (function (_super) {
    tslib_1.__extends(BioblocksVisualization, _super);
    function BioblocksVisualization(props) {
        var _this = _super.call(this, props) || this;
        _this.bioblocksHooks = {};
        _this.datasets = immutable_1.Set();
        _this.loadingStatus = 2 /* NOT_STARTED */;
        BioblocksVisualization.activeBioblocksVisualizations = BioblocksVisualization.activeBioblocksVisualizations.add(_this);
        _this.setupDataServices();
        return _this;
    }
    BioblocksVisualization.prototype.componentWillUnmount = function () {
        this.teardown();
    };
    BioblocksVisualization.prototype.addBioblocksHook = function (name, cb) {
        this.bioblocksHooks[name] = cb;
    };
    BioblocksVisualization.prototype.registerDataset = function (name, defaultValue, namespace) {
        if (namespace === void 0) { namespace = 'bioblocks'; }
        this.datasets = this.datasets.add(new data_1.Dataset(name, namespace));
        this.createReducer(name, defaultValue, namespace);
    };
    BioblocksVisualization.prototype.finishLoading = function () {
        this.loadingStatus = 3 /* READY */;
    };
    BioblocksVisualization.prototype.getDatasets = function () {
        return this.datasets;
    };
    BioblocksVisualization.prototype.getComponentBioblocksHooks = function () {
        return this.bioblocksHooks;
    };
    BioblocksVisualization.prototype.getLoadingStatus = function () {
        return this.loadingStatus;
    };
    BioblocksVisualization.prototype.teardown = function () {
        // TODO Can it be enforced that this method is always called?
        BioblocksVisualization.activeBioblocksVisualizations = BioblocksVisualization.activeBioblocksVisualizations.remove(this);
    };
    BioblocksVisualization.prototype.createReducer = function (datasetName, defaultValue, namespace) {
        if (namespace === void 0) { namespace = 'bioblocks'; }
        if (util_1.isArray(defaultValue)) {
            reducer_1.createContainerReducer(datasetName, namespace);
        }
        else if (typeof defaultValue === 'object') {
            reducer_1.createObjectReducer(datasetName, namespace);
        }
        else {
            reducer_1.createValueReducer(datasetName, defaultValue, namespace);
        }
    };
    BioblocksVisualization.getActiveBioblocksVisualizations = function () {
        return BioblocksVisualization.activeBioblocksVisualizations;
    };
    BioblocksVisualization.activeBioblocksVisualizations = immutable_1.Set();
    return BioblocksVisualization;
}(React.Component));
exports.BioblocksVisualization = BioblocksVisualization;
//# sourceMappingURL=BioblocksVisualization.js.map