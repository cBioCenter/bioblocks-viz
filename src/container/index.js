"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./BioblocksVisualization"), exports);
tslib_1.__exportStar(require("./AnatomogramContainer"), exports);
tslib_1.__exportStar(require("./InfoPanelContainer"), exports);
tslib_1.__exportStar(require("./NGLContainer"), exports);
tslib_1.__exportStar(require("./ContactMapContainer"), exports);
tslib_1.__exportStar(require("./ProteinFeatureViewer"), exports);
tslib_1.__exportStar(require("./SpringContainer"), exports);
tslib_1.__exportStar(require("./TensorTContainer"), exports);
tslib_1.__exportStar(require("./UMAPSequenceContainer"), exports);
tslib_1.__exportStar(require("./UMAPTranscriptionalContainer"), exports);
//# sourceMappingURL=index.js.map