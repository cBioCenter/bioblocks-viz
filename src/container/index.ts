export * from './IGVContainer';
export * from './ProteinFeatureViewer';
export * from './SpringContainer';
export * from './VizPanelContainer';
