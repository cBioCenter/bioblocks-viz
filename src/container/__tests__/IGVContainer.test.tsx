import { shallow } from 'enzyme';
import * as React from 'react';

import { IGVContainer } from '~chell-viz~/container';

describe('IGVContainer', () => {
  it('Should match the snapshot when using default props.', () => {
    const wrapper = shallow(<IGVContainer />);
    expect(wrapper).toMatchSnapshot();
  });
});
