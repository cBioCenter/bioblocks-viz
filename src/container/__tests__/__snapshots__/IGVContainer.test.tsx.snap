// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IGVContainer Should match the snapshot when using default props. 1`] = `
<div
  className="igv-container"
  style={
    Object {
      "height": "600px",
      "width": "600px",
    }
  }
/>
`;
